### ISSUE TITLE
[c10d] bfloat16 NCCL PREMUL_SUM silently zeroes the reduction for power-of-two factors

### ISSUE THREAD

**Issue by bogatyy** [pytorch/pytorch#190745]
## Bug description

`ReduceOp.PREMUL_SUM(factor)` over a bfloat16 tensor multiplies by an incorrect factor: for a host (Python float) scalar whose low 2 bytes are zero, which includes every power-of-two value, the reduction comes back all zeros. This is what FSDP2 hits: with a custom `gradient_divide_factor`, `_get_gradient_divide_factors` issues `ReduceOp.PREMUL_SUM(1 / factor)` over bf16 gradients (`torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py`), so a factor of 2, 4, 8, ... produces exact-zero gradients.

## Repro (2 GPUs)

```python
# torchrun --standalone --nproc_per_node=2 repro.py
import torch, torch.distributed as dist, os
dist.init_process_group("nccl")
torch.cuda.set_device(int(os.environ["LOCAL_RANK"]))
rank, ws = dist.get_rank(), dist.get_world_size()

for dtype in (torch.float16, torch.float32, torch.bfloat16):
    v = torch.full((8,), rank + 1.0, device="cuda", dtype=dtype)
    dist.all_reduce(v, op=dist.ReduceOp.PREMUL_SUM(0.5))
    expected = 0.5 * sum(i + 1 for i in range(ws))
    print(dtype, "got", v.float().mean().item(), "expected", expected)
```

On 2 ranks (NCCL >= 2.11.1):

```
torch.float16  got 1.5  expected 1.5
torch.float32  got 1.5  expected 1.5
torch.bfloat16 got 0.0  expected 1.5   <-- wrong
```

Verified failing on torch 2.9.1+cu128 and a 2.12 nightly on 2x H100. bf16 device-tensor factors are separately rejected with a dtype error.

## Root cause

`getNcclReduceOp` in `torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp` dispatches the bfloat16 case through `unpackPreMulSum<float, ncclBfloat16>`. `ncclRedOpCreatePreMulSum` reads `ncclTypeSize(dataType)` bytes from the factor pointer, i.e. 2 bytes for `ncclBfloat16`, but the factor is stored as a 4-byte `float scalar_factor`. On little-endian hardware NCCL reads the low 2 bytes; for any float whose low half is zero (all powers of two, e.g. 0.5 = 0x3F000000, low bytes 0x0000) it receives a bfloat16 zero. The `half` and `float` cases are correct because there the scalar type width matches the reduction datatype. The device-factor path fails differently: `const_data_ptr<float>()` requires a float tensor, so a bf16 factor tensor raises.

## Fix

Use `at::BFloat16` as the scalar type for the bfloat16 case (matching the `at::Half` / `float` cases), so the factor is the correct 2-byte width for both the host-immediate and device residences. One line in `getNcclReduceOp`, plus extending the PreMulSum coverage in `test/distributed/test_c10d_ops_nccl.py` (currently `half, float, double`) to include `bfloat16`.

## Reachability

Requires FSDP2 with a custom `gradient_divide_factor` that differs from the reduce-scatter group size, and bf16 `reduce_dtype` (the FSDP2 default takes the AVG path and is safe). This is the expert-parallel / MoE gradient-scaling pattern (e.g. PrimeRL sets `set_gradient_divide_factor` on expert FSDP units). Direct `ReduceOp.PREMUL_SUM(host_scalar)` on any bf16 collective is affected regardless of FSDP2.

## Prior art

Distinct from #155903 / #155904 / #155915 (PreMulSum tensor-factor plumbing). No issue found for the bf16 host-scalar width bug.

*This report was authored with assistance from Claude; the repro and root cause were human-reviewed. Found in a correctness audit alongside a parallel independent audit.*

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh
